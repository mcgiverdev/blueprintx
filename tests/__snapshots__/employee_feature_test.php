<?php

declare(strict_types=1);

namespace Tests\Feature;

use App\Modules\Hr\Domain\Models\Department;
use App\Modules\Hr\Domain\Models\Employee;
use App\Modules\Hr\Infrastructure\Seeders\EmployeeSeeder;
use Illuminate\Foundation\Testing\RefreshDatabase;
use Illuminate\Foundation\Testing\WithFaker;
use Tests\TestCase;

class EmployeeFeatureTest extends TestCase
{
    use RefreshDatabase;
    use WithFaker;

    protected function setUp(): void
    {
        parent::setUp();

        $this->seed(EmployeeSeeder::class);
    }

    public function test_can_list_employees(): void
    {
        $response = $this->getJson('/hr/employees');

        $response->assertOk();
    }

    public function test_can_show_employee(): void
    {
        $employee = Employee::factory()->create();

        $response = $this->getJson("/hr/employees/{$employee->id}");

        $response->assertOk();
        $response->assertJsonPath('data.id', (string) $employee->id);
    }

    public function test_can_create_employee(): void
    {
        $department = Department::factory()->create();

        $payload = [
            'first_name' => $this->faker->firstName(),
            'last_name' => $this->faker->lastName(),
            'email' => $this->faker->unique()->safeEmail(),
            'salary' => $this->faker->randomFloat(2, 30000, 150000),
            'active' => true,
            'department_id' => (string) $department->id,
        ];

        $response = $this->postJson('/hr/employees', $payload);

        $response->assertCreated();
        $response->assertJsonPath('data.first_name', $payload['first_name']);
    }

    public function test_can_update_employee(): void
    {
        $employee = Employee::factory()->create();

        $payload = [
            'first_name' => 'Updated First Name',
            'last_name' => 'Updated Last Name',
            'email' => 'updated.email@example.com',
            'salary' => '42000.00',
            'active' => false,
        ];

        $response = $this->putJson("/hr/employees/{$employee->id}", $payload);

        $response->assertOk();
        $response->assertJsonPath('data.first_name', $payload['first_name']);
    }

    public function test_can_delete_employee(): void
    {
        $employee = Employee::factory()->create();

        $response = $this->deleteJson("/hr/employees/{$employee->id}");

        $response->assertNoContent();
        $this->assertSoftDeleted(Employee::class, ['id' => $employee->id]);
    }

    public function test_can_toggle_active_flag(): void
    {
        $employee = Employee::factory()->active(false)->create();

        $response = $this->patchJson("/hr/employees/{$employee->id}/toggle-active");

        $response->assertOk();
        $response->assertJsonPath('data.active', true);
    }

    public function test_can_update_salary(): void
    {
        $employee = Employee::factory()->create();

        $payload = [
            'salary' => '85000.00',
        ];

        $response = $this->patchJson("/hr/employees/{$employee->id}/update-salary", $payload);

        $response->assertOk();
        $response->assertJsonPath('data.salary', $payload['salary']);
    }

    public function test_can_search_employees(): void
    {
        Employee::factory()->create([
            'first_name' => 'Jane',
            'last_name' => 'Doe',
            'email' => 'jane.doe@example.com',
        ]);

        $response = $this->getJson('/hr/employees?search=Jane');

        $response->assertOk();
        $response->assertJsonCount(1, 'data');
        $response->assertSeeText('Jane', $escaped = false);
    }

    public function test_can_get_active_stats(): void
    {
        Employee::factory()->active()->create();
        Employee::factory()->active(false)->create();

        $response = $this->getJson('/hr/employees/stats/active');

        $response->assertOk();
        $response->assertJsonStructure([
            'data' => [
                ['active', 'count'],
            ],
        ]);
    }
}
