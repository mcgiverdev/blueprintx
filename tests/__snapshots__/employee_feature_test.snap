<?php

namespace Tests\Feature\Hr;

use Illuminate\Foundation\Testing\RefreshDatabase;
use Tests\TestCase;
use App\Domain\Hr\Models\Employee;
use App\Domain\Hr\Models\Department;

class EmployeeFeatureTest extends TestCase
{
    use RefreshDatabase;

    public function test_index_returns_paginated_list(): void
    {
        $department = Department::factory()->create();
        Employee::factory()->count(3)->create([
            'department_id' => $department->id,
        ]);

        $response = $this->getJson('/api/hr/employees');

        $response
            ->assertOk()
            ->assertJsonStructure([
                'data',
                'meta' => ['current_page', 'from', 'last_page', 'per_page', 'to', 'total'],
            ]);

        $this->assertSame(3, $response->json('meta.total'));
        $this->assertNotNull($response->json('data.0.department.id'));
        $this->assertSame($department->id, $response->json('data.0.department.id'));
    }

    public function test_can_store_employee(): void
    {
        $department = Department::factory()->create();

        $payload = [
            'first_name' => 'Sample First Name',
            'last_name' => 'Sample Last Name',
            'email' => 'john.doe@example.com',
            'salary' => '4200.00',
            'active' => true,
            'department_id' => $department->id,
            'hired_at' => '2025-01-01T10:00:00Z',
        ];

        $response = $this->postJson('/api/hr/employees', $payload);

        $response->assertCreated();
        $response->assertJsonFragment([
            'first_name' => 'Sample First Name',
            'last_name' => 'Sample Last Name',
        ]);
        $response->assertHeader('ETag');
        $this->assertNotNull($response->json('data.department.id'));
        $this->assertSame($department->id, $response->json('data.department.id'));

        $this->assertDatabaseHas('employees', [
            'first_name' => 'Sample First Name',
            'last_name' => 'Sample Last Name',
            'email' => 'john.doe@example.com',
            'salary' => '4200.00',
            'active' => true,
            'department_id' => $department->id,
            'hired_at' => '2025-01-01T10:00:00Z',
        ]);
    }

    public function test_can_update_employee(): void
    {
        $department = Department::factory()->create();

        $employee = Employee::factory()->create([
            'department_id' => $department->id,
        ]);

        $payload = [
            'first_name' => 'Updated First Name',
            'last_name' => 'Updated Last Name',
            'email' => 'jane.doe@example.com',
            'salary' => '6500.50',
            'active' => false,
            'department_id' => $department->id,
            'hired_at' => '2025-12-31T15:30:00Z',
        ];

        $response = $this->putJson('/api/hr/employees/' . $employee->id, $payload, [
            'If-Match' => '*',
        ]);

        $response->assertOk();
        $response->assertJsonFragment([
            'first_name' => 'Updated First Name',
            'last_name' => 'Updated Last Name',
        ]);
        $response->assertHeader('ETag');
        $this->assertNotNull($response->json('data.department.id'));
        $this->assertSame($department->id, $response->json('data.department.id'));

        $this->assertDatabaseHas('employees', [
            'id' => $employee->id,
            'first_name' => 'Updated First Name',
            'last_name' => 'Updated Last Name',
            'email' => 'jane.doe@example.com',
            'salary' => '6500.50',
            'active' => false,
            'department_id' => $department->id,
            'hired_at' => '2025-12-31T15:30:00Z',
        ]);
    }

    public function test_can_delete_employee(): void
    {
        $department = Department::factory()->create();

        $employee = Employee::factory()->create([
            'department_id' => $department->id,
        ]);

        $response = $this->deleteJson('/api/hr/employees/' . $employee->id, [], [
            'If-Match' => '*',
        ]);

        $response->assertNoContent();

        $this->assertSoftDeleted('employees', [
            'id' => $employee->id,
        ]);
    }
}
